numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
     Py_RETURN_NONE;
 }
 
+static void array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    // This will also free the strides as it's one allocation.
+    PyMem_Free(self->dl_tensor.shape);
+    PyMem_Free(self);
+    Py_XDECREF(array);
+}
+
+
+NPY_NO_EXPORT PyObject *
+from_dlpack(PyObject *NPY_UNUSED(self), PyObject *obj) {
+    PyObject *capsule = PyObject_CallMethod(obj, "__dlpack__", NULL);
+    if (capsule == NULL) {
+        return NULL;
+    }
```

Comment:
This isn't how dunder lookup usually works; does `dlpack` specify these semantics?

The rule for all builtin dunders (`_Py_LookupSpecial`) is roughly `type(obj).__dict__[dunder](obj)`, which bypasses metaclasses and the object dictionary.